hyperspy/io_plugins/semper_unf.py

```diff
@@ -0,0 +1,520 @@
+            # Read title:
+            title = ''
+            if ntitle > 0:
+                assert np.frombuffer(f.read(4), dtype=np.int32)[0] == ntitle  # length of title
+                title_bytes = np.frombuffer(f.read(ntitle), dtype=np.byte)
+                title = ''.join(map(chr, title_bytes))
+                assert np.frombuffer(f.read(4), dtype=np.int32)[0] == ntitle
+            # Read label:
+            iwp, date, range_string, ipltyp = [None] * 4  # Initialization!
+            iccoln, icrown, iclayn = [None] * 3
+            x0, y0, z0 = 0., 0., 0.
+            dx, dy, dz = 1., 1., 1.
+            ux, uy, uz = '', '', ''
+            if ilabel:
+                rec_length = np.frombuffer(f.read(4), dtype=np.int32)[0]  # length of label
+                label = np.frombuffer(f.read(512), dtype=np.int16)
+                assert ''.join([chr(l) for l in label[:6]]) == 'Semper'
+                assert struct.unpack('>h', ''.join([chr(x) for x in label[6:8]]))[0] == ncol
+                assert struct.unpack('>h', ''.join([chr(x) for x in label[8:10]]))[0] == nrow
+                assert struct.unpack('>h', ''.join([chr(x) for x in label[10:12]]))[0] == nlay
+                iccoln = struct.unpack('>h', ''.join([chr(x) for x in label[12:14]]))[0]
+                icrown = struct.unpack('>h', ''.join([chr(x) for x in label[14:16]]))[0]
+                iclayn = struct.unpack('>h', ''.join([chr(x) for x in label[16:18]]))[0]
+                assert label[18] == iclass
+                assert label[19] == iform
+                iwp = label[20]
+                date = '{}-{}-{} {}:{}:{}'.format(label[21]+1900, *label[22:27])
+                # No test for ncrang, range is extracted from data itself (also prone to errors)!
+                ipltyp = label[55]  # position list type
+                real_coords = label[62]
+                dz, dy, dx, z0, y0, x0 = [1., 1., 1., 0., 0., 0.]
+                if real_coords:
+                    dz = struct.unpack('<f', ''.join([chr(x) for x in label[75:79]]))[0]
+                    z0 = struct.unpack('<f', ''.join([chr(x) for x in label[79:83]]))[0]
+                    dy = struct.unpack('<f', ''.join([chr(x) for x in label[83:87]]))[0]
+                    y0 = struct.unpack('<f', ''.join([chr(x) for x in label[87:91]]))[0]
+                    dx = struct.unpack('<f', ''.join([chr(x) for x in label[91:95]]))[0]
+                    x0 = struct.unpack('<f', ''.join([chr(x) for x in label[95:99]]))[0]
+                assert ''.join([str(unichr(l)) for l in label[100:100+ntitle]]) == title
+                ux = ''.join([chr(l) for l in label[244:248]]).replace('\x00', '')
+                uy = ''.join([chr(l) for l in label[248:252]]).replace('\x00', '')
+                uz = ''.join([chr(l) for l in label[252:256]]).replace('\x00', '')
+                assert np.frombuffer(f.read(4), dtype=np.int32)[0] == rec_length

```

Comment:
I would move all the "label extraction" to a separate (private) function to increase code readability. That could also include a nice checking that the it is actually formatted properly (i.e. all the `assert`s), and if not, just return an empty label and a warning (so that you can still load the data and only label is lost)